File: DSA/src/Questions/Array/MergeTwoSortedArray.java
package Questions.Array;

public class MergeTwoSortedArray {
    public static void merge(int[] nums1, int m, int[] nums2, int n) {
        int final_size = nums1.length - 1;
        m--;
        n--;
        while (m > -1 && n > -1) {
            if (nums1[m] < nums2[n]) {
                int temp = nums1[final_size];
                nums1[final_size--] = nums2[n];
                nums2[n--] = temp;
            } else {
                int temp = nums1[final_size];
                nums1[final_size--] = nums1[m];
                nums1[m--] = temp;
            }
        }
        for (int k = 0; k < nums1.length; k++) {
            System.out.print(nums1[k] + " ");
        }
        System.out.println("");
        for (int k = 0; k < nums2.length; k++) {
            System.out.print(nums2[k] + " ");
        }
        while (n > -1) {
            int temp = nums1[final_size];
            nums1[final_size--] = nums2[n];
            nums2[n--] = temp;
        }
    }

    public static void main(String[] args) {
        int[] nums1 = new int[] { 1, 2, 3, 0, 0, 0 };
        int[] nums2 = new int[] { 2, 5, 6 };
        merge(nums1, 3, nums2, 3);
    }
}

File: DSA/Python/Questions/Array/MergeTwoSortedArray.py
"""
# Question: Merge Two Sorted Arrays
# Merge two sorted arrays in-place with optimal space usage

# Time Complexity: O(m + n) where m and n are lengths of input arrays
# Space Complexity: O(1) as we modify arrays in-place

# Algorithm:
# 1. Start from end of both arrays
# 2. Compare and swap elements to maintain sorted order
# 3. Handle remaining elements from second array
# 4. Use final_size pointer for placement

# Key Components:
# - merge(): Main function implementing in-place merge
# - Three-pointer technique with swapping
"""

from typing import List


class MergeTwoSortedArray:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        final_size = len(nums1) - 1
        m -= 1
        n -= 1

        while m > -1 and n > -1:
            if nums1[m] < nums2[n]:
                nums1[final_size], nums2[n] = nums2[n], nums1[final_size]
                n -= 1
            else:
                nums1[final_size], nums1[m] = nums1[m], nums1[final_size]
                m -= 1
            final_size -= 1

        while n > -1:
            nums1[final_size], nums2[n] = nums2[n], nums1[final_size]
            n -= 1
            final_size -= 1


def main():
    solution = MergeTwoSortedArray()
    nums1 = [1, 2, 3, 0, 0, 0]
    nums2 = [2, 5, 6]
    solution.merge(nums1, 3, nums2, 3)
    print(nums1)  # Expected output: [1, 2, 2, 3, 5, 6]


if __name__ == "__main__":
    main()
